### ISSUE TITLE
`ClientSideBaseVisitor.fragmentsGraph` is recomputed several times despite never changing

### ISSUE THREAD

**Issue by vhfmag** [dotansimha/graphql-code-generator#10018]
### Which packages are impacted by your issue?

@graphql-codegen/visitor-plugin-common

### Describe the bug

While profiling the graphql codegen step of one of my company's projects a sharp difference in build time after migrating from `near-operation-file` to `client-preset`, we identified that a disproportionate time is spent in [these three lines](https://github.com/vhfmag/graphql-code-generator/blob/1b8656fa3577e5568af4ce5d6aadbbb6384658fb/packages/plugins/other/visitor-plugin-common/src/client-side-base-visitor.ts#L389-L391):

```js
const fragmentDependencyNames = new Set(
  fragmentNames.map(name => this.fragmentsGraph.dependenciesOf(name)).flatMap(item => item)
);
```

You can see that in the flamegraph below, where all of the green traces on the bottom are time spent on those lines — I added a couple arrows to help visualize which trace I mean, but it's all traces of the same color, not just these 4

<img width="1849" alt="Screenshot 2024-06-21 at 15 51 28" src="https://github.com/dotansimha/graphql-code-generator/assets/357835/34884032-03cd-4372-96b4-d660b8faaf47">

The reason why that line is so expensive is that `this.fragmentsGraph` is [an expensive getter](https://github.com/dotansimha/graphql-code-generator/blob/master/packages/plugins/other/visitor-plugin-common/src/client-side-base-visitor.ts#L522-L549). In addition to that, it only depends on `this._fragments` and `this._extractFragments()`. The `this._extractFragments` only depends on `this._fragments` and itself (it's recursive), and `this._fragments` isn't reassigned or modified (which we verified by changing its declaration to `private readonly _fragments: ReadonlyMap<string, LoadedFragment>`).

With that in mind, we converted `fragmentsGraph` from a getter to a regular property, and our build time went from ~130 seconds to ~50s. I'm opening this issue to share our findings and share a PR proposing a related performance improvement.

### Your Example Website or App

N/A

### Steps to Reproduce the Bug or Issue

1. Run codegen on a mid-sized project using `client-preset` and measure the build time
2. Make the change to only compute `fragmentsGraph` once per `ClientSideBaseVisitor` instance
3. Run codegen again and compare the build time

### Expected behavior

`client-preset` should be faster by avoiding costly recomputations of unchanged data

### Platform

- OS: macOS
- NodeJS: 18.18.2
- `graphql` version: 16.2.0
- `@graphql-codegen/client-preset` version(s): 4.2.6
- `@graphql-codegen/visitor-plugin-common` version(s): 2.13.1

**Comment by vhfmag** [comment]
Hi @n1ru4l! I see you merged #10019 (thanks!), but there hasn't been a version release yet. I'm wondering what the usual release process is

**Comment by n1ru4l** [comment]
@vhfmag Releasing it right now 😇 https://github.com/dotansimha/graphql-code-generator/pull/10010
